### PR TITLE
Jira V3 merge candidate

### PR DESCRIPTION
**Pull request by levb**

### CODE REVIEW

File: server/http.go [hunk 1]
```diff
@@ -59,67 +78,102 @@ func (p *Plugin) ServeHTTP(c *plugin.Context, w http.ResponseWriter, r *http.Req
...
 
 func (p *Plugin) serveHTTP(c *plugin.Context, w http.ResponseWriter, r *http.Request) (int, error) {
-	switch r.URL.Path {
+	var err error
+	path := r.URL.Path
+	instanceURL := ""
+	instanceURL, path = splitInstancePath(path)
+	callbackInstanceID := types.ID("")
+
+	// Add any "callback" URLs here so that the caller instance is properly identified
```
Comment (levb) [inline server/http.go:87]: Need a better comment: that _some_ routes expect the instance ID prefix.

File: server/http.go [hunk 2]
```diff
@@ -196,6 +250,7 @@ func httpWorkflowRegister(p *Plugin, w http.ResponseWriter, r *http.Request) (in
 						Description: "Jira issue ID",
 					},
 				},
+				//TODO <><> prefix route with instance? or unnecessary since it's for all instances?
```
Comment (levb) [inline server/http.go:253]: Resolve the TODO

File: server/utils/telemetry/rudder.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
+package telemetry
+
+import rudder "github.com/rudderlabs/analytics-go"
+
+const (
+	rudderDataPlaneURL = "https://pdat.matterlytics.com"
+	rudderWriteKey     = "1d5bMvdrfWClLxgK1FvV3s4U1tg"
```
Comment (levb) [inline server/utils/telemetry/rudder.go:7]: @larkox We need to change this, right?

File: server/user.go [hunk 4]
```diff
@@ -67,14 +73,20 @@ func httpUserConnect(ji Instance, w http.ResponseWriter, r *http.Request) (int,
...
 
+	instance, err := p.instanceStore.LoadInstance(instanceID)
+	if err != nil {
+		return respondErr(w, http.StatusInternalServerError, err)
+	}
+
 	// Users shouldn't be able to make multiple connections.
-	jiraUser, err := ji.GetPlugin().userStore.LoadJIRAUser(ji, mattermostUserId)
-	if err == nil && len(jiraUser.Key()) != 0 {
+	// TODO <> this block needs to be updated. Though idk if this route will still get called?
```
[the remark this answers is not in the document]
Comment (jfrerich) [reply server/user.go:82]: https://github.com/mattermost/mattermost-plugin-jira/issues/611

File: server/http.go (hunk 2 above)
Comment (jfrerich) [reply server/http.go:253]: I created an issue to track and resolve all TODO comments.  https://github.com/mattermost/mattermost-plugin-jira/issues/611

File: server/command.go [hunk 5]
```diff
@@ -1,49 +1,295 @@
...
+	"github.com/mattermost/mattermost-plugin-jira/server/utils/types"
 )
 
+const commandTrigger = "jira"
+
+var jiraCommandHandler = CommandHandler{
+	handlers: map[string]CommandHandlerFunc{
+		"assign":                  executeAssign,
+		"connect":                 executeConnect,
+		"debug/clean-kv":          executeDebugCleanKV,
```
[the remark this answers is not in the document]
Comment (jfrerich) [reply server/command.go]: Done

File: server/plugin.go [hunk 6]
```diff
@@ -119,7 +123,17 @@ type Plugin struct {
...
-	webhookQueue chan []byte
+	webhookQueue chan *webhookMessage
+
+	// telemetry client
+	telemetryClient telemetry.Client
+
+	// telemetry Tracker
+	Tracker jiraTracker.Tracker
+
+	// service that determiines if this Mattermost instance has access to
```
[the remark this answers is not in the document]
Comment (jfrerich) [reply server/plugin.go]: Done

File: server/http.go (hunk 1 above)
Comment (jfrerich) [reply server/http.go:87]: I'm trying to rethink the wording here, but I think this comment makes enough sense as is.

File: server/utils/telemetry/rudder.go (hunk 3 above)
Comment (larkox) [reply server/utils/telemetry/rudder.go:7]: If we want to use the library in experimental, you can follow directly the instructions [there](https://github.com/mattermost/mattermost-plugin-api/blob/master/experimental/telemetry/doc.go). If we were to keep this implementation, the only difference with the documentaion would be to set the right path to the `LDFLAGS` part.

The makefile may not use LDFLAGS. In that case, you can take a look on how it is done on the PRs for [Zoom](https://github.com/mattermost/mattermost-plugin-zoom/pull/148) or [Todo](https://github.com/mattermost/mattermost-plugin-todo/pull/101).

**Comment by DHaussermann** [comment]
Thanks @levb I did alot of testing on this PR and this looks very good!
I did not end to end test all the Jira v2 functionality but, I did retest everything new to V3 and did some smoke testing of the most common use cases for the existing functionality (create, attach, Webhooks, user DM etc...)

Testing included:
- Regression tested migration for Cloud and Server  
- Regression tested that only E20 can add a 2nd or more instances 
- Verified Create, attach and subscribe UI when additional instances are installed but not connected
- Verify no --instance needed when only one instance is connected 
- Verified use of slash command default instance when additional instances are installed but not connected
- Connecting to additional instance
- Subscribe list work without issue when legacy instance is included
- Spot checked various slash commands with --instance being passed in and not relying on default  
- Verified use of slash command default instance when additional instances are connected
- Verified Create, attach and subscribe UI when additional instances are connected
- Verified default instance selected as expected in modal
- Verified instance specific default project working in Create and Attach modal
- Added subscription from various Jira instances and confirmed expected behavior 
- Jira Group permission ignore non legacy instance
- Updates to default instance correctly reflected in slash command when no instance is passed 
- Subscribe shows only connected instances
- User can still access subscribe as expected for some instance when they don't have permission on other connected instances
- Subscribe edit modal updates dynamically based on permissions 
- For 3+ instances subscribe UI correctly shows connected instances only 
- For 3+ instances Create and Attach modal correctly shows connected instances only 
- For 3+ instances slash command --instance correctly reflects connected instances only   
- Confirmed most recent PRs to address nil legacy instance crash and show project name have been included 

No issues found.
LGTM!

**Comment by DHaussermann** [comment]
cc @aaronrothschild to keep you posted on the progress of this as I know where looking to update the Beta release as soon as we can.

**Pull request merged**